c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -229,12 +349,55 @@ fn assign_origins<'tcx>(
         } else {
             hypothesis[lty.label]
         };
+
+        let construct_adt_origins = |ty: &Ty, amaps: &mut AtomMaps| -> Option<&_> {
+            let adt_def = ty.ty_adt_def()?;
+
+            // create a concrete origin for each actual or hypothetical
+            // lifetime parameter in this ADT
+            let origins: Vec<_> = adt_metadata
+                .table
+                .get(&adt_def.did())?
+                .lifetime_params
+                .iter()
+                .map(|o| {
+                    let pairing = (*o, amaps.origin());
+                    eprintln!("pairing lifetime parameter with origin: {pairing:?}");
+                    pairing
+                })
+                .collect();
+
+            if origins.is_empty() {
+                return None;
+            }
+
+            Some(ltcx.arena().alloc_slice(&origins[..]))
```

Comment:
```suggestion
            let origin_iter = adt_metadata
                .table
                .get(&adt_def.did())?
                .lifetime_params
                .iter()
                .map(|o| {
                    let pairing = (*o, amaps.origin());
                    eprintln!("pairing lifetime parameter with origin: {pairing:?}");
                    pairing
                });
            let origins: &[_] = ltcx.arena().alloc_from_iter(origin_iter);
            Some(origins).filter(|origins| !origins.is_empty())
```

Reply:
`alloc_from_iter` I think is a good change - the iterator is an `ExactSizeIterator`, so it should be able to preallocate the right amount of space in the arena.

I'm not a fan of `Some(x).filter(...)`, but actually I wonder if the `Option` is needed here at all.  Currently the `Label::origin_params` field has two different "empty" values: `None` and `Some(&[])`.  Is it actually necessary to distinguish between those cases, or could we remove the `Option` and just have `&[]` as the only "empty" state?  (I also wonder if `origin` and `origin_params` could be folded into a single `&[(OriginParam, Origin)]` field for uniformity, though this would need a new `OriginParam` variant for the built-in origin parameter of `TyKind::Ref` and `TyKind::RawPtr`.)

Reply:
address in 34bc309f except for consolidating `origin` and `origin_param`; that's a larger change involving new code for `Aggregate` and `Cast` handling, and I'd like to do that in another PR.